modules/thermalobject.c

```diff
@@ -0,0 +1,679 @@
+{
+    const char *res = NULL;
+    if( thermal_object_has_valid_sensor_vtab(self) )
+        res = self->to_sensor_vtab->tsv_get_depends_on_cb(self);
+    return res;
+}
+
+/** Get the currently cache stable status of thermal object
+ *
+ * @param self  thermal object pointer
+ *
+ * @return thermal status, or THERMAL_STATUS_INVALID
+ */
+THERMAL_STATUS
+thermal_object_get_status(const thermal_object_t *self)
+{
+    return self ? self->to_status_curr : THERMAL_STATUS_INVALID;
+}
+
+/** Get the currently temperature value of thermal object
+ *
+ * @param self  thermal object pointer
+ *
+ * @return temperature, or INVALID_TEMPERATURE
+ */
+int
+thermal_object_get_temperature(const thermal_object_t *self)
+{
+    return self ? self->to_temperature : INVALID_TEMPERATURE;
+}
+
+/** Get the polling delay for thermal object
+ *
+ * The polling delay is defined by the sensor backend and depend
+ * on the sensor backend status.
+ *
+ * @param self  thermal object pointer
+ *
+ * @return true if mintime/maxtime was filled in, false otherwise
+ */
+bool
+thermal_object_get_poll_delay(thermal_object_t *self,
+                              int *mintime, int *maxtime)
+{
+    bool ack = false;
+
+    if( !thermal_object_has_valid_sensor_vtab(self) )
+        goto EXIT;
+
+    ack = self->to_sensor_vtab->tsv_get_poll_delay_cb(self, mintime, maxtime);
+
+EXIT:
+    return ack;
+}
+
+/** Check if the thermal object is about to change status
+ *
+ * @param self  thermal object pointer
+ *
+ * @return true if thermal status is unstable, false otherwise
+ */
+bool
+thermal_object_status_in_transition(const thermal_object_t *self)
+{
+    return self && self->to_status_change_started > 0;
+}
+
+/** [Optional] Helper for logging thermal object status changes
+ *
+ * @param self  thermal object pointer
+ */
+#if DSME_THERMAL_LOGGING
+static void
+thermal_object_log_status(const thermal_object_t *self)
+{
+    static const char  log_path[]  = "/var/lib/dsme/thermal.log";
+    static FILE       *log_file    = 0;
+    static bool        did_open  = false;
+
+    struct timespec t;
+    clock_gettime(CLOCK_MONOTONIC, &t);
+
+    int now = t.tv_sec;
+    static int start_time = 0;
+
+    if( !start_time ) {
+        start_time = now;
+    }
+    int temperature = thermal_object_get_temperature(self);
+
+    dsme_log(LOG_DEBUG, PFIX"%s: %dC %s", thermal_object_get_name(self),
+             temperature, thermal_status_name(self->to_status_curr));
+
+    if( !did_open ) {
+        did_open = true;
+
+        if( !(log_file = fopen(log_path, "a")) ) {
+            dsme_log(LOG_ERR, PFIX"%s: Error opening thermal log: %m",
+                     log_path);
+        }
+    }
+
+    if( log_file ) {
+        fprintf(log_file, "%d %s %d C %s\n",
+                now - start_time,
+                thermal_object_get_name(self),
+                temperature,
+                thermal_status_name(self->to_status_curr));
+        fflush(log_file);
```

Comment:
Interesting observation. This particular function comes from the old implementation, so I did not pay too much attention to it apart from making mechanical changes.

You are right: The file should be closed already when thermal manager plugin is unloaded.

However, since it is just a debugging feature that is not enabled by default, implicit close at dsme exit is good enough (plugins are unloaded only when dsme is about to make exit).